batchflow/utils_transforms.py

```diff
@@ -0,0 +1,179 @@
+
+    def normalize(self, array, normalization_stats=None, return_stats=False):
+        """ Normalize image with provided stats.
+
+        Parameters
+        ----------
+        array : numpy.ndarray
+
+        normalization_stats : dict or None, optional
+            If provided, then used to get statistics for normalization, by default None.
+            If None, self.normalization_stats will be used.
+            If self.normalization_stats is also None, then statistics will by computed by array.
+        return_stats : bool, optional
+            Whether to return stats used for normalization, by default False
+
+        Returns
+        -------
+        numpy.ndarray or (numpy.ndarray, dict)
+        """
+        clip_to_quantiles = self.clip_to_quantiles
+        array = array if self.inplace else array.copy()
+
+        if normalization_stats is None:
+            if self.normalization_stats is not None:
+                normalization_stats = self.normalization_stats
+            else:
+                if clip_to_quantiles:
+                    np.clip(array, *np.quantile(array, self.q), out=array)
+                    clip_to_quantiles = False
+
+                if callable(self.mode):
+                    normalization_stats = {
+                        'mean': np.mean(array),
+                        'std': np.std(array),
+                        'min': np.min(array),
+                        'max': np.max(array),
+                    }
+                else:
+                    normalization_stats = {}
+                    if 'mean' in self.mode:
+                        normalization_stats['mean'] = np.mean(array)
+                    if 'std' in self.mode:
+                        normalization_stats['std'] = np.std(array)
+                    if 'min' in self.mode:
+                        normalization_stats['min'] = np.min(array)
+                    if 'max' in self.mode:
+                        normalization_stats['max'] = np.max(array)
+        if clip_to_quantiles:
+            np.clip(array, normalization_stats['q_01'], normalization_stats['q_99'], out=array)
+
+        # Actual normalization
+        if callable(self.mode):
+            array[:] = self.mode(array, normalization_stats)
+        else:
+            if 'mean' in self.mode:
+                array -= normalization_stats['mean']
+            if 'std' in self.mode:
+                array /= normalization_stats['std'] + 1e-6
+            if 'min' in self.mode and 'max' in self.mode:
+                if clip_to_quantiles:
+                    array -= normalization_stats['q_01']
+                    array /= normalization_stats['q_99'] - normalization_stats['q_01']
+                elif normalization_stats['max'] != normalization_stats['min']:
+                    array -= normalization_stats['min']
+                    array /= normalization_stats['max'] - normalization_stats['min']
+                else:
+                    array -= normalization_stats['min']
```

Comment:
What do you think about this way of parsing parts of string `mode` VS hardcoded options (`meanstd`, `minmax`, maybe `mean`)?

Reply:
I like the current scheme